analyzer/library/bundler/bundler.go

```diff
@@ -2,48 +2,41 @@ package bundler
 
 import (
 	"bytes"
+	"os"
 	"path/filepath"
 
-	"github.com/aquasecurity/fanal/types"
-
 	"github.com/aquasecurity/fanal/analyzer/library"
 
 	"github.com/aquasecurity/fanal/analyzer"
-	"github.com/aquasecurity/fanal/extractor"
 	"github.com/aquasecurity/fanal/utils"
 	"github.com/aquasecurity/go-dep-parser/pkg/bundler"
-	godeptypes "github.com/aquasecurity/go-dep-parser/pkg/types"
 	"golang.org/x/xerrors"
 )
 
 func init() {
-	analyzer.RegisterLibraryAnalyzer(&bundlerLibraryAnalyzer{})
+	analyzer.RegisterAnalyzer(&bundlerLibraryAnalyzer{})
 }
 
+var (
+	requiredFiles = []string{"Gemfile.lock"}
+)
+
 type bundlerLibraryAnalyzer struct{}
 
-func (a bundlerLibraryAnalyzer) Analyze(fileMap extractor.FileMap) (map[types.FilePath][]godeptypes.Library, error) {
-	libMap := map[types.FilePath][]godeptypes.Library{}
-	requiredFiles := a.RequiredFiles()
-
-	for filename, content := range fileMap {
-		basename := filepath.Base(filename)
-		if !utils.StringInSlice(basename, requiredFiles) {
-			continue
-		}
-
-		r := bytes.NewBuffer(content)
-		libs, err := bundler.Parse(r)
-		if err != nil {
-			return nil, xerrors.Errorf("error with %s: %w", filename, err)
-		}
-		libMap[types.FilePath(filename)] = libs
+func (a bundlerLibraryAnalyzer) Analyze(content []byte) (analyzer.AnalyzeReturn, error) {
+	r := bytes.NewBuffer(content)
+	libs, err := bundler.Parse(r)
+	if err != nil {
+		return analyzer.AnalyzeReturn{}, xerrors.Errorf("error with Gemfile.lock: %w", err)
 	}
-	return libMap, nil
+	return analyzer.AnalyzeReturn{
+		Libraries: libs,
+	}, nil
 }
 
-func (a bundlerLibraryAnalyzer) RequiredFiles() []string {
-	return []string{"Gemfile.lock"}
+func (a bundlerLibraryAnalyzer) Required(filePath string, _ os.FileInfo) bool {
+	fileName := filepath.Base(filePath)
+	return utils.StringInSlice(fileName, requiredFiles)
```

Comment:
A general observation between the various library/*.go files is that there's a lot of duplicate logic especially for functions like `Analyze`, `Required`. Could we DRY this up somehow? If we end up doing so it will also become easier to write a tests (less duplicate setups to keep).

Reply:
I thought it was flexible to implement in each library/*.go, but I'll try DRY.

Reply:
I tried the following, but I feel like it doesn't help so much. The original function is small enough, so we can't reduce so many lines. Do you have any idea?

```
func analyze(content []byte, parse func(r io.Reader) ([]godeptypes.Library, error)) (analyzer.AnalyzeReturn, error) {
	r := bytes.NewBuffer(content)
	libs, err := parse(r)
	if err != nil {
		return analyzer.AnalyzeReturn{}, xerrors.Errorf("error with a lock file: %w", err)
	}
	return analyzer.AnalyzeReturn{
		Libraries: libs,
	}, nil
}

type bundlerLibraryAnalyzer struct{}

func (a bundlerLibraryAnalyzer) Analyze(content []byte) (analyzer.AnalyzeReturn, error) {
	ret, err := analyze(content, bundler.Parse)
	if err != nil {
		return analyzer.AnalyzeReturn{}, xerrors.Errorf("unable to parse Gemfile.lock: %w", err)
	}
	return ret, nil
}
```

Reply:
I like this idea. I pushed some changes here https://github.com/aquasecurity/fanal/pull/110/commits/0f897b6e4727fe9fd216f2a050f6d4a3455a9d0b